Feature/530 group model error handling

PROgram52bc opened the pull request:
The `Make invalid request` button should be removed before merge, but currently it provides a way to see the effect of the error displayed.

Commit:
Fixed error handling 404

- add more detail on friendly 404 error message

PROgram52bc committed:
implemented experimental error snackbar

- Currently, when the content of the message is too much, some of it
will not be displayed; I think this issue will be less obvious with a
newer version of vuetify, though.
- The snackbar currently does not support displaying multiple messages
at once; new messages will replace existing messages, without resetting
the timer for the previous messages.

PROgram52bc committed:
reverted 404 error messages to id

Since an invalid id does not correspond to a name, reporting the id is
probably the best we can do to report the error

PROgram52bc committed:
added standard mechanism to log API activity

PROgram52bc committed:
added logging statement to api

PROgram52bc committed:
added 'new request' info to group apis

this is just an experimentation; we might remove them in the future.
But the point is that we can log 'normal' information to the console
without logging them into the log file.

Commit:
Add translation functionality to snackbar

- snackbar able to switch between English and Spanish

- Add more tests for ha

Commit:
Update snackbar translation

- swithch between English and Spanish

Merge branch 'feature/530-group-model-error-handeling' of https://github.com/corpus-christi/corpus-christi into feature/530-group-model-error-handling

PROgram52bc committed:
add test jwt to database

in order to make sure test tokens generated are not recognized as
revoked tokens

PROgram52bc committed:
updated api log functionality

- added a 'log' helper function to log the response body and status code
- changed the api endpoints to use the 'log' helper function instead of
calling logger directly
- modified logger's config to include url, remote_addr, and various
endpoint-related information

PROgram52bc committed:
added log files to gitignore

PROgram52bc committed:
reverted more api error response to refer to id

PROgram52bc committed:
updated logging

- added BASE_DIR in the log file path
- disabled logging in tests
- 'log' function accept a single response object

Commit:
Add report message card

- report error message from user

Commit:
Add error card

- user can send error report through this card
Merge branch 'feature/530-group-model-error-handeling' of https://github.com/corpus-christi/corpus-christi into feature/530-group-model-error-handling

nurkkala commented on api/src/groups/api.py:
```diff
@@ -24,29 +24,29 @@ def create_group_type():
     db.session.add(new_group_type)
     db.session.commit()
-    return jsonify(group_type_schema.dump(new_group_type)), 201
+    return log(jsonify(group_type_schema.dump(new_group_type)), 201)
```
Looking through this file, the `log(jsonify())` pattern appears a _lot_. Could we define a new function (maybe `logged` or `logged_response`) that simply took two arguments? The function could inspect the first argument; if it is an object, then invoke `jsonify`.

PROgram52bc replied:
Sure. That makes sense. Should we replace `log` with that maybe?

nurkkala replied:
Yes.

nurkkala commented on api/src/shared/helpers.py:
```diff
@@ -114,6 +116,51 @@ def parse_kv_str(kv_str):
+        if isinstance(args[0], tuple):
+            body, code = args[0]
+        else: 
+            body, code = args[0], 200
```
I like this default.

nurkkala commented on api/src/shared/helpers.py:
```diff
@@ -114,6 +116,51 @@ def parse_kv_str(kv_str):
+        the second argument is the status code
+        
+    calling with 1 argument:
+        the single argument is a 2-element tuple:
```
How often do you think we'll pass a tuple?

PROgram52bc replied:
I think that's the case where we return directly from `modify_entity`, which returns a tuple, and wrapped with the `log` function, we have a tuple being passed to the `log` function. Maybe we can destructure it, like `log(*modify_entity(...))`?

PROgram52bc replied:
Actually, `modify_entity` will return an already `jsonified` Response, which can't be passed to the `logged_response` because it will try to `jsonify` it again. Should I try to inspect whether the object passed in is already `jsonified` in `logged_response`, or should I simply implement the update without `modify_entity`?

nurkkala commented on ui/i18n/cc-i18n.json:
```diff
@@ -34,12 +34,14 @@
       "login": "Log In",
       "logout": "Log Out",
       "open-map": "Open Map",
+      "Report-error": "Report Error(not implemented)",
```
Let's keep the format consistent with all lower-case letters.

nurkkala commented on ui/i18n/cc-i18n.json:
```diff
@@ -805,12 +810,14 @@
       "login": "Acceder",
       "logout": "Cerrar sesi\u00f3n",
       "open-map": "Abrir Mapa",
+      "Report-error": "Informar error (no implementado)",
```
Lower case

quiteqiang replied:
solved

nurkkala commented on ui/i18n/cc-i18n.json:
```diff
@@ -1016,6 +1023,9 @@
       "error-archiving-event": "Error al archivar el evento",
       "error-editing-event": "Error al editar evento",
       "error-unarchiving-event": "Error al desarchivar el evento",
+      "error-result-conflect": "Conflicto",
```
Misspelling -- should be `conflict` instead of `conflect`.

nurkkala commented on ui/src/components/MessageSnackBar.vue:
```diff
@@ -0,0 +1,96 @@
+  <div>
+    <!-- TODO: add multiple messages support? 
+      Currently new messages will override the old ones, and count down will not restart <2020-06-04, David Deng> -->
+    <!-- START: to be removed -->
```
Do we want to commit this example?

PROgram52bc replied:
Probably not yet. There are still things to be finalized. Maybe I will create a new branch without the changes in the UI and merge those first?

nurkkala commented on ui/src/components/MessageSnackBar.vue:
```diff
@@ -0,0 +1,96 @@
+      multi-line
+      top
+    >
+      <blog-post v-if= "snackBarObj.content == '404'">{{ $t("events.error-result-not-found") }}</blog-post>
```
What is a `blog-post` control? Is this a new Vuetify thing?

PROgram52bc replied:
I will review these code again... Maybe close this PR for now, and we will make a new one when it is more ready (I just wanted some of your feedback on the logging functionality).

nurkkala commented on ui/src/components/MessageSnackBar.vue:
```diff
@@ -0,0 +1,96 @@
+    -->
+  <div>
+    <!-- TODO: add multiple messages support? 
+      Currently new messages will override the old ones, and count down will not restart <2020-06-04, David Deng> -->
```
Good question. My initial guess is that this won't happen to often, so it may not be worth a ton of extra effort right now.

nurkkala reviewed (commented):
Great work. Almost there ...

PROgram52bc committed:
update logging functionality

- renamed log to logged_response, and simplified the functionality by
accepting only objects that have not be 'jsonified'.
- changed the update endpoints to modify attributes without using
'modify_entity' function

PROgram52bc committed:
Merge branch 'feature/530-group-model-error-handeling' of github.com:corpus-christi/corpus-christi into feature/530-group-model-error-handeling

Commit:
Fixed grammar

PROgram52bc commented:
@quiteqiang updated the grammar in e8a56e3, and relevant changes in the logging functionality are in 0dcb088.
A new PR will be submitted once the UI side is closer to a finishing state :)
